Assignment2 - ChangGyo

Dunkkkk opened the pull request:
# 구조

## Constant & Common
카드에 정적정보와 CardState enum을 구현하고 있다.

## MainGameBoardView
카드에 모든 뷰들을 통합하고 EnvironmentObject를 통해 게임이 완료되면 보여줄 뷰를 구성한다.
해당 뷰에서 onAppear시 theme을 바꿔주어서, 기존에 빈값으로 초기화 되어있던 정보들을 새로 할당해준다.


## view
### CardView
각 카드를 보여주는 view
onTapGesture를 통해 viewmodel에 intent를 전달하고,
cardstate를 통해 앞면 뒷면 맞춘카드를 표시하여 해당 되는 ui를 보여준다.

### TitleView
EnvironmentObject 로부터 현재 테마와 Point를 받아 보여준다

### CardListView
CardView를 스크롤 그리드형태로 보여주고, 해당 테마의 색으로 바꿔준다.

### BottomButtonView
새로운 게임을 시작할 수 있는 NextGame 버튼을 보여준다.

## ViewModel
### MemorizeGameManger
두가지 model을 published형태 즉 관찰가능한 형태로 두었고,
테마변경, 카드선택 이라는 두 가지 함수를 두어서 model내부에서 해당 값이 변경하도록 하였다.

## Model
### Theme
테마 정보를 저장하는 model로써 테마를 변경하는 하나의 함수를 가지고 있다.

#### KindOf
테마의 종류를 나타내는 열거형으로써, Theme의 중첩된 형태로 선언되었다.

### imojiGame
테마의 변경에 따른 카드를 새로 할당하고, 섞는 changeGameByTheme 함수와
카드를 선택했을 때 뒤집어주는 toggleCard 함수
카드의 결과를 포인트에 더하고 빼주는 calculatePoint 함수가 있다.

#### Card
카드에 대한 정보와, imoji, id등을 가지고 있는 중첩된 열거형이다.


# 부족하다고 생각되는 점
## 1. model이 너무 과한 것 같다.
코드를 작성하다보니 model에 너무 많은 로직이 있는것 같다.
핵심적인 로직이라긴 보단 모든 로직이 다 들어가 있는 듯한 느낌이 들어
new architecture라는 곳에 새로운 viewmodel과 model을 작성중인데
시간상의 이유로 우선 PR하였다.

new architecture에선 model에는 정말 기본적인 toggle만을 남기고 viewmodel에서 처리하는 것을 설계하여 구현중이다.

## 2. imojiGame의 계산 로직이 너무 복잡한 것 같다.
충분히 간단하고 예외가 더 적은 계산로직을 짤 수 있을 것 같다.

## 3. commit하는 습관
우선 이번 과제는 인터넷이 되지 않은 곳에서 대부분 진행하여서 commit을 하지 못했지만, 실제로 commit을 하는 단위가 너무 큰 것 같다.

[영상이 커서 유투브로 올렸습니다!!](https://www.youtube.com/shorts/OvkOsTTq9ms)

Dunkkkk committed:
make MemorizeGame with MVVM

Details will be written in PR message.

gen-com commented on week2/2022-iOS-2week/ContentView.swift:
```diff
@@ -0,0 +1 @@
+
```
불필요한 파일은 없애주세요.

gen-com commented on week2/2022-iOS-2week/LifeCycle.swift:
```diff
@@ -0,0 +1,13 @@
+import SwiftUI
+
+@main
+struct changgyoAssignment: App {
```
struct의 이름은 대문자여야 합니다.

Dunkkkk replied:
수정했습니다!

gen-com commented on week2/2022-iOS-2week/LifeCycle.swift:
```diff
@@ -0,0 +1,13 @@
+
+@main
+struct changgyoAssignment: App {
+    let environmentObject = MemorizeGameManger()
```
environmentObject 네이밍이 최선인가요 ?

let, 즉 상수로 전달했지만 실제로 변경이 가능한 이유는 무엇일까요 ?

Dunkkkk replied:
memorizeGameManager가 클래스이기 때문에 즉 참조값 주소가 저장되기때문에
그 주소자체가 바뀌거나 새롭게 어떤것이 할당 된다면 안되겠지만
주소 안에 들어가 그 내부에 프로퍼티들이 변경되는 것은 가능하기 때문입니다

gen-com commented on week2/2022-iOS-2week/MainGameBoardView.swift:
```diff
@@ -0,0 +1,31 @@
+                    Spacer()
+                    Text("Finish this round")
+                        .font(.system(size: 40))
+                    Text("you got a " + "\(gameManager.gameManage.point)" + " Point")
```
`+` 로 나눈 이유는 무엇인가요 ? 그냥 쓰면 되지 않나요 ?

Text("you got a \(gameManager.gameManage.point) Point")
영어 문법이 좀 .. you got 9 points, you got a point(1점을 받은 경우)

Dunkkkk replied:
죄송합니다..... 영어 공부하겠습니다 ㅠㅠ

gen-com commented on week2/2022-iOS-2week/MainGameBoardView.swift:
```diff
@@ -0,0 +1,31 @@
+            }
+            BottomButtonList()
+        }
+        .onAppear(){
```
onAppear를 사용하는 이유는 무엇인가요 ?

Dunkkkk replied:
테마 모델과 게임 모델을 서로 나눌려고 노력했고, 테마가 초기화 되고 게임을 초기화해야하는데 지연실행으로 초기화를 시키는 것은 안 좋은 것같아서 비여있는 인스턴스로 초기화시키고, onappear를 사용하여 그위에 테마로 덮어주었습니다.
사실 처음부터 특정 한 값을 기본 게임으로 등록시키면 되지만, 처음 실행됬을때도 랜덤으로 게임이 구성되기를 원했기 때문에 위처럼 했습니다

gen-com commented on week2/2022-iOS-2week/SubView/PlusMinusButton.swift:
```diff
@@ -0,0 +1,18 @@
+import SwiftUI
+
+struct PlusMinusButton: View {
+   
```
아마 사용하지 않는 파일 같지만 그래도 공백이 통일되지 않네요

Dunkkkk replied:
삭제 하였습니다

Dunkkkk committed:
Remove unnecessary files and Fix by PRreviews

Dunkkkk committed:
Change some variable name and fix grammar

rose6649 commented on week2/2022-iOS-2week/Constant&Common/ConstantCardContent.swift:
```diff
@@ -0,0 +1,26 @@
+
+struct ConstantImojiContent {
```
Imoji가 맞는 스펠링일까요???
아이모지가 된거같은데 !
그릐고 앞에 Constant를 네이밍으로 사용하는 이유를 알 수 있을까요??

Dunkkkk replied:
스펠링 유념하겠습니다 .
정적인 이모지가 있어서 그렇게 네이밍 한것 같습니다 수정했습니다

rose6649 commented on week2/2022-iOS-2week/viewmodel/MemorizeGameManger.swift:
```diff
@@ -0,0 +1,16 @@
+
+import SwiftUI
+
+class MemorizeGameManger: ObservableObject{
```
띄어쓰기

Dunkkkk replied:
수정 하였습니다~

rose6649 commented on week2/2022-iOS-2week/viewmodel/MemorizeGameManger.swift:
```diff
@@ -0,0 +1,16 @@
+
+class MemorizeGameManger: ObservableObject{
+    @Published var themeModel: Theme = Theme()
+    @Published var gameManage: ImojiGame = ImojiGame()
```
gameManage라는 네이밍이 약간 어법을 잘 모르겠고 뭘 의미하는지 더 명확했으면 좋겠어요!
MemorizeGameManager.gameManage 면 뭔가 네이밍이 중복적이지 않나요?

Dunkkkk replied:
수정 했습니다

rose6649 commented on week2/2022-iOS-2week/Constant&Common/CommonEnum.swift:
```diff
@@ -0,0 +1,17 @@
+
+enum CardState {
+    case isFaceUp
+    case isFaceDown
+    case isMatched
+    
+    mutating func toggle() {
+        switch self {
+        case .isFaceUp:
+            self = .isFaceDown
+        case .isFaceDown:
+            self = .isFaceUp
+        case .isMatched:
+            self = .isMatched
+        }
+    }
+}
```
CardState는 Card에 종속적인것 같은데,
따로 파일이 나와있는 이유를 알 수 있을까요??

Dunkkkk replied:
수정 했습니다

rose6649 commented on week2/2022-iOS-2week/Constant&Common/ConstantCardContent.swift:
```diff
@@ -0,0 +1,26 @@
+    static let vehicle = ["🚗","🚑","🚙","🚓","🚒","🚕","🛺","🚐","🚌","🛻","🚎","🚚","🏎","🚛","🏍","🛵","🚁"]
+    static let outfit = ["👚","🥼","🦺","🧥","👙","🩱","👞","🥿","🥾","🧢","👡","👠","🧦","👘","👗","🧤","🧣","👓"]
+    
+    static func emitImoji(willChangeTheme theme: Theme.KindOf) -> [String] {
+        switch theme {
+        case .face:
+            return face
+        case .animal:
+            return animal
+        case .sports:
+            return sports
+        case .vehicle:
+            return vehicle
+        case .heart:
+            return heart
+        case .outfit:
+            return outfit
+        }
+    }
```
혹시 Theme 구조체에 없고 ConstantImojiContent를 따로 둔 이유를 알 수 있을까요?

Dunkkkk replied:
원래 제네릭이 아니였어서 Emojigame에 넣어두었어야하는데 밖으로 분리했습니다. 이유는 한 구조체가 길어지는게 뭔가 잘못된거 같았는데 다른 멘토님께 리뷰 받고 제네릭으로 수정하고, 제네릭안에 static을 둘 수 없어서 다시 밖으로 뺐습니다

rose6649 commented on week2/2022-iOS-2week/View/CardView.swift:
```diff
@@ -0,0 +1,39 @@
+    var body: some View {
+        ZStack {
+            let shape = RoundedRectangle(cornerRadius: 20)
+            switch content.isFaceUp {
+            case .isFaceUp:
+                shape
+                    .fill()
+                    .foregroundColor(.white)
+                shape
+                    .strokeBorder(lineWidth: 5)
+                VStack {
+                    Text(content.imoji)
+                        .font(.largeTitle)
+                }
+            case .isFaceDown:
+                shape
+                    .fill()
+            case .isMatched:
+                shape
+                    .fill()
+                    .foregroundColor(.white)
+            }
+        }
```
지금 switch가 content.isFaceUp.isFaceUp, .isFaceUp.isFaceDown...
이 되고 있는거 같아요.
한번만 확인해주세요!

Dunkkkk replied:
cardstate를 통해 수정완료했습니다

rose6649 commented on week2/2022-iOS-2week/model/ImojiGame.swift:
```diff
@@ -0,0 +1,83 @@
+    }
+    
+    mutating private func calculatePoint() {
+        if cards[openCards[0]].imoji == cards[openCards[1]].imoji {
+            point += 3
+            completeCardSet += 1
+            if completeCardSet == cards.count / 2 {
+                shouldFinishThisRound = true
+            }
+        }
+        else if !cards[openCards[0]].didSelected && !cards[openCards[1]].didSelected {
+            point += 0
+        }
+        else if cards[openCards[0]].didSelected && cards[openCards[1]].didSelected {
+            point -= 2
+        }
+        else {
+            point -= 1
+        }
```
배열의 직접접근은 피해주세용!

Dunkkkk replied:
수정완료했습니다

rose6649 commented on week2/2022-iOS-2week/model/ImojiGame.swift:
```diff
@@ -0,0 +1,83 @@
+extension ImojiGame {
+    struct Card: Identifiable, Hashable {
+        var didSelected = false
+        var isFaceUp: CardState = .isFaceDown
```
여기서 뭔가 변수명이 달라진거였군요

Dunkkkk replied:
수정 완료했습니다! 넵 ㅠㅠ

Dunkkkk committed:
Fix changeGameLogic and faceUpCardPointCalculate function and Card struct is changed generic

fix changeGameLogic by using static and change some logic in EmojiGame and Card Struct is changed generic

leejun6694 commented on week2/2022-iOS-2week/ChanggyoAssignment.swift:
```diff
@@ -0,0 +1,14 @@
+
+@main
+struct ChanggyoAssignment: App {
+    static let initGame = MemorizeGameManger.startNewGame()
```
static 인 이유가 있나요~?

leejun6694 commented on week2/2022-iOS-2week/Model/EmojiGame.swift:
```diff
@@ -0,0 +1,108 @@
+
+struct EmojiGame<CardContentType: Equatable> {
+    var leftCardSetCount: Int { cards.count - cards.filter { $0.state == .isMatched }.count }
```
줄바꿈~!
```
var leftCardSetCount: Int { 
   cards.count - cards.filter { $0.state == .isMatched }.count
}
```

leejun6694 commented on week2/2022-iOS-2week/Model/EmojiGame.swift:
```diff
@@ -0,0 +1,108 @@
+        let endIndex = startIndex + displayCardCount
+        var newCards = [Card]()
+        
+        for i in startIndex ... endIndex {
```
줄여서 사용하지 않습니다~!
`for index in `

leejun6694 commented on week2/2022-iOS-2week/Model/EmojiGame.swift:
```diff
@@ -0,0 +1,108 @@
+    mutating func changeStateCard(cardId: Int) {
+        guard let chosenIndex = cards.firstIndex(where: {cardId == $0.id}) else { return }
+        
+        if cards[chosenIndex].state == .isMatched { return }
```
```
guard let chosenIndex = cards.firstIndex(where: { cardId == $0.id })
   || cards[chosenIndex].state != .isMatched else {
    return
}
```
요렇게 묶을 수도 있어요~!

leejun6694 commented on week2/2022-iOS-2week/Model/EmojiGame.swift:
```diff
@@ -0,0 +1,108 @@
+    
+    mutating private func calculatePoint() {
+        guard let firstCardIndex = cards.firstIndex(where: { $0.state == .isFaceUp }),
+            let secondCardIndex = cards.lastIndex(where: { $0.state == .isFaceUp }) else { return }
```
index 를 활용할거면, index 들을 모아놓고 거기서 first, last 를 사용하면 어떨까요?
두번 전체 돌리는 것보단?

leejun6694 commented on week2/2022-iOS-2week/Model/EmojiGame.swift:
```diff
@@ -0,0 +1,108 @@
+        guard let firstCardIndex = cards.firstIndex(where: { $0.state == .isFaceUp }),
+            let secondCardIndex = cards.lastIndex(where: { $0.state == .isFaceUp }) else { return }
+
+        if cards[firstCardIndex].Emoji == cards[secondCardIndex].Emoji {
```
`cards[firstCardIndex]`, `cards[secondCardIndex]`
두 값이 계속 사용되는데, 변수로 선언해주고 사용하면 어떨까요~?

leejun6694 commented on week2/2022-iOS-2week/Model/Theme.swift:
```diff
@@ -0,0 +1,62 @@
+        case heart
+        case outfit
+        
+        var name: String {
```
rawValue 로 사용하는건 어떨까요?

```
var name: String {
   self.rawValue
}
```

rose6649 commented on week2/2022-iOS-2week/Model/EmojiGame.swift:
```diff
@@ -0,0 +1,108 @@
+        if cards[firstCardIndex].Emoji == cards[secondCardIndex].Emoji {
+            point += 3
+            if leftCardSetCount == 2 {
+                cards[firstCardIndex].state = .isMatched
```
카드 state있는게 보기 편하네용 ㅎㅎ

rose6649 commented on week2/2022-iOS-2week/Viewmodel/MemorizeGameManger.swift:
```diff
@@ -0,0 +1,38 @@
+    @Published private(set) var gameModel: EmojiGame<String>
+    @Published private(set) var finishThisRound = false
+    
+    init(initGame: (themeModel: Theme, gameModel: EmojiGame<String>)) {
+        self.themeModel = initGame.themeModel
+        self.gameModel = initGame.gameModel
+    }
```
themeModel과 gameModel을 분리시켜 같은 레벨의 다른 모델로 생성한 이유가 있을까요? :)

Dunkkkk committed:
make Calculator with SwiftUI

use two model ( Calculate, CalculatorButton ) with leat function
many operand excution are decalred in Calculator

Dunkkkk committed:
add model

Dunkkkk committed:
make Calculator with jumin

Dunkkkk committed:
combine Project with jumin

fix some error ( back to back operation, click equal in succession, etc.. )
change name ( Buttonview -> NumberPad, CalculatorButtonGrid -> NumberPadGrid )

Dunkkkk committed:
add slide to move and some string extension

gen-com commented on week4/Calculator/Calculator/CalculatorApp.swift:
```diff
@@ -0,0 +1,20 @@
+@main
+struct CalculatorApp: App {
+  
+  let caculatorManager = CalculatorManager()
```
caculatorManager -> calculatorManager (오타)

접근제어 !

gen-com commented on week4/Calculator/Calculator/Extensions/Ex+Color.swift:
```diff
@@ -0,0 +1,22 @@
+
+extension Color {
+  
+  init(hex: String) {
```
해당 코드에 대해 설명해주세요.

구글링으로 가져온 코드라면 주석으로 링크를 걸어주세요.

gen-com commented on week4/Calculator/Calculator/Extensions/Ex+Color.swift:
```diff
@@ -0,0 +1,22 @@
+
+import SwiftUI
+
+extension Color {
```
앱에서 사용되는 고정적인 색은 전역상수로 둔다면 어떤가요 ?

gen-com commented on week4/Calculator/Calculator/CalculatorApp.swift:
```diff
@@ -0,0 +1,20 @@
+import SwiftUI
+
+@main
+struct CalculatorApp: App {
```
여기서는 들여쓰기가 통일되지 않은 것 같네요. 2칸 ? 4칸 ?

gen-com commented on week4/Calculator/Calculator/Extensions/Ex+String.swift:
```diff
@@ -0,0 +1,17 @@
+
+extension String {
+  
+  mutating func removeWithDecimal() {
```
가장 작은 자리를 없애는 메서드 인것 같네요. 네이밍은 좀 개선되었으면 합니다.

gen-com commented on week4/Calculator/Calculator/Model/CalculatorModel.swift:
```diff
@@ -0,0 +1,218 @@
+//
+
+import Foundation
+
```
전반적으로 코드가 읽히지 않아서 로직을 이해하기 힘들었습니다.

그리고 아이폰 계산기에 대한 분석을 모두 고려하지는 않은 것 같습니다.

gen-com commented on week4/Calculator/Calculator/Model/CalculatorModel.swift:
```diff
@@ -0,0 +1,218 @@
+    }
+  }
+  
+  enum InputState {
```
Input보다는 계산에 필요한 연산자, Operand 가 적절하지 않을까요 ?

gen-com replied:
Input보다는 계산에 필요한 피연산자, Operand 가 적절하지 않을까요 ?

아래의 상태들이 무엇인지 알기 어렵습니다.
new(새로운), previous(이전) -> 새로우면서 이전인 상태는 어떤거죠 ?
혹시 첫 피연산자가 필요한 상태라면, firstOperandNeeded와 같은 표현은 어떤가요 ?
아래도 모두 마찬가지 입니다.

gen-com commented on week4/Calculator/Calculator/Model/CalculatorModel.swift:
```diff
@@ -0,0 +1,218 @@
+struct Calculator {
+  
+  var state: InputState = .newPreviousNumber
+  var operation: ArithmeticOperation? {
```
이미 사용되고 있는 키워드를 변수명으로 지정하고 싶을때는 아래와 같이 할 수 있습니다.

``` swift
var `operator`: ArithmeticOperation?
```

gen-com commented on week4/Calculator/Calculator/Model/CalculatorModel.swift:
```diff
@@ -0,0 +1,218 @@
+    }
+  }
+  
+  mutating func joinNumbers(with nextElement: Decimal) -> Decimal {
```
joinNumbers 는 외부에서 접근할 필요가 없을 것 같은데요 ? 접근 제어를 잊으셨나요 ?
``` swift
private mutating func joinNumber(_ number: Decimal) -> Decimal {
    Decimal(string: showingText + "\(number)") ?? 0
}
```
이렇게도 할 수 있을 거 같네요

gen-com commented on week4/Calculator/Calculator/Model/CalculatorModel.swift:
```diff
@@ -0,0 +1,218 @@
+    }
+  }
+  
+  enum clearType {
```
enum 이름은 대문자로 시작합니다.

gen-com commented on week4/Calculator/Calculator/Model/CalculatorModel.swift:
```diff
@@ -0,0 +1,218 @@
+        return false
+      }
+    }
+    
```
앞서 말했듯, 자주 사용하는 색은 Color의 extension에서 전역상수로 두는 것을 고려해보세요.

gen-com commented on week4/Calculator/Calculator/Model/CalculatorModel.swift:
```diff
@@ -0,0 +1,218 @@
+        return "."
+      }
+    }
+    
```
모델에 View의 요소가 있습니다.
MVVM 을 준수하기 위해 구조를 다시 고려해 보세요.

View 내부적으로 작동되는 로직은 View로 빼보세요.
Model에 의해 View가 영향을 받는다면, 그에 대한 해석은 ViewModel에 맞겨보세요.

여러 연산이나 숫자를 enum으로 모델에 정의하는 것은 좋습니다만, 이로 비롯되는 것은 모델의 역할을 넘는 것 아닐까요 ?

gen-com commented on week4/Calculator/Calculator/View/ContentView.swift:
```diff
@@ -0,0 +1,34 @@
+  
+  var body: some View {
+    ZStack{
+      Color.black.ignoresSafeArea()
```
SafeArea는 무엇인가요 ?

gen-com commented on week4/Calculator/Calculator/View/ContentView.swift:
```diff
@@ -0,0 +1,34 @@
+
+import SwiftUI
+
+struct ContentView: View {
```
네이밍을 확실히 바꿔야 하겠죠 ?

CalculatorView ?

gen-com commented on week4/Calculator/Calculator/View/ContentView.swift:
```diff
@@ -0,0 +1,34 @@
+      VStack {
+        Spacer()
+        ResultView(showingText: calculatorManager.showingText)
+          .gesture(
```
```swift
/// Gesture Handling
    private func swipeRight() -> some Gesture {
        /// 최소 30 정도로 이동했다면, 그리고 오른쪽(양수) 방향이라면, 로직 수행
        DragGesture(minimumDistance: 30, coordinateSpace: .local)
            .onEnded { drag in
                if 0 < drag.translation.width {
                    calculatorManager.slideToRemove()
                }
            }
    }
```
위와 같이 제스처에 대해 메서드를 정의하고
```swift
ResultView(showingText: calculatorManager.showingText)
    .gesture(swipeRight())
```
이와 같이 쓸 수 있습니다.

translation은 무엇인지 알아볼까요 ?

gen-com commented on week4/Calculator/Calculator/View/ContentView.swift:
```diff
@@ -0,0 +1,34 @@
+            DragGesture()
+              .onEnded { _ in calculatorManager.slideToRemove() }
+          )
+        NumberPadGrid(items: calculatorManager.buttonList , columnsCount: 4) { item in
```
상수 처리 하면 좋지 않을까요 ?

띄어쓰기가 좀 이상하지 않나요 ?

gen-com commented on week4/Calculator/Calculator/View/SubViews/NumberPad.swift:
```diff
@@ -0,0 +1,41 @@
+
+import SwiftUI
+
+struct NumberPad: View {
```
Pad 보다는 Button을 구성하는 거 같은데 맞나요 ?

Button을 그리드 형식으로 모아 둔 것을 패드라고 하지 않나요 ?

gen-com commented on week4/Calculator/Calculator/View/SubViews/NumberPad.swift:
```diff
@@ -0,0 +1,41 @@
+
+extension NumberPad {
+  
+  private enum DrawConstans {
```
상수 처리는 언제나 좋습니다.
Constans -> Constants(오타 주의)

gen-com commented on week4/Calculator/Calculator/View/ContentView.swift:
```diff
@@ -0,0 +1,34 @@
+  @ObservedObject var calculatorManager: CalculatorManager
+  
+  var body: some View {
+    ZStack{
```
띄어쓰기

gen-com commented on week4/Calculator/Calculator/View/SubViews/NumberPadGrid.swift:
```diff
@@ -0,0 +1,89 @@
+
+import SwiftUI
+
+struct NumberPadGrid<buttonListView: View>: View {
```
Pad와 Grid는 다소 중복의 의미가 있는 것 같습니다.

buttonListView, 제너릭스도 타입으로 대문자로 써야 합니다. ButtonListView

gen-com commented on week4/Calculator/Calculator/View/SubViews/NumberPadGrid.swift:
```diff
@@ -0,0 +1,89 @@
+  let columnsCount: Int
+  let content: (Calculator.ArithmeticOperation) -> buttonListView
+  
+  var body: some View {
```
상수 처리 !

gen-com commented on week4/Calculator/Calculator/View/SubViews/NumberPadGrid.swift:
```diff
@@ -0,0 +1,89 @@
+    }
+  }
+  
+  init(items: [Calculator.ArithmeticOperation], columnsCount: Int, @ViewBuilder content: @escaping (Calculator.ArithmeticOperation) -> buttonListView) {
```
``` swift
init(
    items: [Calculation.ArithmeticOperation],
    columnsCount: Int,
    @ViewBuilder content: @escaping (Calculation.ArithmeticOperation) -> buttonListView
) {
    buttonDataArray = ButtonDataArray.makeTwoDemsionArrayForColumnsCount(
        items: items, 
        columnsCount: columnsCount
    )
    self.columnsCount = columnsCount
    self.content = content
}
```
파라미터가 길다면 위처럼 들여쓰기 하면 깔끔해 집니다.
불필요한 self는 빼는 것을 지향해 주세요.

gen-com commented on week4/Calculator/Calculator/View/SubViews/NumberPadGrid.swift:
```diff
@@ -0,0 +1,89 @@
+
+extension NumberPadGrid {
+  
+  struct ButtonDataArray: Identifiable {
```
아래의 코드를 설명해주세요.

rose6649 replied:
ButtonDataArray라는 이름이 조금 어색하지 않나요? 죽내려서 아래 코드 보면 Array인데 또 배열안에 들어가네요
그리고 DataArray같은 상징적 단어는 좀 지양하는게 좋을 것 같아요 :)
다른 네이밍도 한 번 생각해봐주세요!

gen-com commented on week4/Calculator/Calculator/View/SubViews/NumberPadGrid.swift:
```diff
@@ -0,0 +1,89 @@
+      array.removeAll()
+    }
+    
+    static func makeTwoDemsionArrayForColumnsCount(items: [Calculator.ArithmeticOperation], columnsCount: Int) -> [ButtonDataArray] {
```
Demsion -> Dimension (오타주의)

gen-com commented on week4/Calculator/Calculator/View/SubViews/ResultView.swift:
```diff
@@ -0,0 +1,35 @@
+    HStack {
+      Spacer()
+      Text(showingText)
+        .lineLimit(1)
```
이 친구도 상수처리 해주는 건 어떤가요 ?

gen-com commented on week4/Calculator/Calculator/View/SubViews/ResultView.swift:
```diff
@@ -0,0 +1,35 @@
+        .minimumScaleFactor(DrawConstans.fontMinimumSacleFactor)
+        .padding(.trailing, DrawConstans.textPadding)
+    }
+    .contentShape(Rectangle())
```
``` swift
.contentShape(Rectangle())
```
가 필요한 이유는 무엇일까요 ?

gen-com commented on week4/Calculator/Calculator/View/SubViews/ResultView.swift:
```diff
@@ -0,0 +1,35 @@
+
+extension ResultView {
+  
+  private enum DrawConstans {
```
DrawConstans -> DrawConstants (오타주의)
아래의 값은 Double이고, 위에서 필요로하는 값은 CGFloat 입니다. 하지만 아무 문제가 일어나지 않는 이유는 무엇일까요 ?

gen-com commented on week4/Calculator/Calculator/ViewModel/CaculatorManager.swift:
```diff
@@ -0,0 +1,114 @@
+          case .plus, .minus, .divide, .multiply:
+            buttonList[index].activeButton(active: false)
+          default:
+            print("error")
```
예외처리는 확실하게 해주세요.

gen-com commented on week4/Calculator/Calculator/ViewModel/CaculatorManager.swift:
```diff
@@ -0,0 +1,114 @@
+      clickOperation(clickType)
+    }
+  }
+  
```
click으로 사용자의 모든 의도를 처리하면, 아래의 메서드들은 접근제어가 필요하지 않을까요 ?

leejun6694 commented on week3/Calculator/Calculator/Model/Calculate.swift:
```diff
@@ -0,0 +1,45 @@
+struct Calculate {
+    var calculateResult: Double = 0
+    var currentNumber: Double = 0
+    var currentOperand: CalculatorButton? = .add(false)
```
초기값을 설정해주는데, optional 로 정의한 이유가 있나요?

leejun6694 commented on week3/Calculator/Calculator/Model/Calculate.swift:
```diff
@@ -0,0 +1,45 @@
+    var currentNumber: Double = 0
+    var currentOperand: CalculatorButton? = .add(false)
+    
+    mutating func calculate(to value: Double) -> calculateError {
```
calculate 계산하는 method 의 return type 이 error 인게 조금 어색합니다.
다르게 정의할 수 있지 않을까요?

leejun6694 commented on week3/Calculator/Calculator/Model/CalculatorButton.swift:
```diff
@@ -0,0 +1,128 @@
+//  Created by changgyo seo on 2022/08/21.
+//
+
+enum CalculatorButton: Identifiable, Equatable {
```
Button 이라고 하면 View 나 UI 를 상속받은 코드로 예상하는데, 내용은 아닌거 같네요...!?

leejun6694 commented on week3/Calculator/Calculator/Model/CalculatorButton.swift:
```diff
@@ -0,0 +1,128 @@
+    var backgroundColor: (red: Double, green: Double, blue: Double, opacity: Double) {
+        switch self {
+        case .clear, .percent, .sign:
+            return (160 / 255, 160 / 255, 160 / 255 , 1)
```
띄어쓰기!

leejun6694 commented on week3/Calculator/Calculator/View/ButtonView.swift:
```diff
@@ -0,0 +1,54 @@
+}
+
+extension ButtonView {
+    private enum DrawConstans {
```
오타!

leejun6694 commented on week4/Calculator/Calculator/Model/CalculatorModel.swift:
```diff
@@ -0,0 +1,218 @@
+  }
+  
+  mutating func changeSign() {
+    if let priorNumber = previousNumber {
```
if let 을 사용해도 되지만, else 가 있는지 기대하게 되네요~
guard 를 사용하는건 어떨까요?

leejun6694 commented on week4/Calculator/Calculator/ViewModel/CaculatorManager.swift:
```diff
@@ -0,0 +1,114 @@
+  
+  var showingText: String {
+    get {
+      if calculateModel.showingText != "0" {
```
코드가 길어지게 되는 경우, method 로 빼는건 어떨까요?

rose6649 commented on week4/Calculator/Calculator/Model/CalculatorModel.swift:
```diff
@@ -0,0 +1,218 @@
+      if nextNumber != nil { calculate() }
+    }
+  }
+  var showingText: String
+  private(set) var previousNumber: Decimal? {
+    didSet {
+      guard let previousNumber = previousNumber else { return }
+      showingText = "\(previousNumber)"
+    }
+  }
```
모델에서 showing Text를 가지는 이유가 있을까요?

rose6649 commented on week4/Calculator/Calculator/View/SubViews/NumberPadGrid.swift:
```diff
@@ -0,0 +1,89 @@
+  struct ButtonDataArray: Identifiable {
+    
+    var id: Int
+    var array: [Calculator.ArithmeticOperation]
```
네이밍이 아쉽네요 ㅠㅠ

rose6649 commented on week4/Calculator/Calculator/View/SubViews/NumberPadGrid.swift:
```diff
@@ -0,0 +1,89 @@
+    
+    static func makeTwoDemsionArrayForColumnsCount(items: [Calculator.ArithmeticOperation], columnsCount: Int) -> [ButtonDataArray] {
+      var arrayWithColumnsCount = ButtonDataArray()
+      var TwoDemesionArray = [ButtonDataArray]()
```
네이밍

rose6649 commented on week4/Calculator/Calculator/ViewModel/CaculatorManager.swift:
```diff
@@ -0,0 +1,114 @@
+class CalculatorManager: ObservableObject {
+  
+  @Published private var calculateModel = create()
+  @Published var buttonList: [Calculator.ArithmeticOperation] =
+  [
+    .clear(.allData), .sign, .percent, .divide(false),
+    .number(7), .number(8), .number(9), .multiply(false),
+    .number(4), .number(5), .number(6), .minus(false),
+    .number(1), .number(2), .number(3), .plus(false),
+    .number(0), .point, .equal
+  ]
```
아까는 DataArray였는데 여기는 네이밍이 List네요~
사실 배열리스트 형태들은 이미 타입유추가 가능해서 네이밍할때 같이 안적기도 해요
buttons 처럼 복수형을 나타내고 더 추가할 구체적 네이밍은 정해서 써도 될것같다는 생각입니다 :D

rose6649 commented on week4/Calculator/Calculator/ViewModel/CaculatorManager.swift:
```diff
@@ -0,0 +1,114 @@
+        if buttonList[index].isActive {
+          switch buttonList[index] {
+          case .plus, .minus, .divide, .multiply:
+            buttonList[index].activeButton(active: false)
```
active가 bool값인것같은데 isActive가 맞지않을까요?

rose6649 commented on week4/Calculator/Calculator/ViewModel/CaculatorManager.swift:
```diff
@@ -0,0 +1,114 @@
+    }
+  }
+  
+  func clickNumber(_ number: Decimal) {
+    calculateModel.clickNumber(number)
+  }
+  
+  func clickOperation(_ operation: Calculator.ArithmeticOperation?) {
+    calculateModel.operation = operation
+    calculateModel.state = .newNextNumber
+  }
```
그냥 참고인데,
func click 오버로딩도 괜찮을거같아요 ! ㅎㅎ

```swift
func click(number: Decimal) {}
func click(opration: Cal..) {}
```
